lib/utils.hpp

```diff
@@ -0,0 +1,34 @@
+// Utilities in the system domain (like logging)
+
+#pragma once
+
+#include <iostream>
+#include <limits>
+#include <string>
+#include <vector>
+
+// Log a string to console
+inline void log(std::string message) {
+    std::cout << message << "\n";
+}
+
+// Log a vector<double> to console
+inline void log(std::vector<double> row) {
+  // log numbers with full precision, usually 17 digits
+  std::cout.precision(std::numeric_limits<double>::max_digits10);
+
+  std::cout << "{";
+  for (double val : row) {
+    std::cout << val << ", ";
+  }
+  std::cout << "}"
+       << "\n";
+}
+
+// Throw exception if index is out of bounds of vector
+template <class T>
```

Comment:
1. merged template function into header
- respects the compile-once principle
- otherwise we cannot specialized vectorGuard, and get undefined errors, eg:
   "vectorGuard<vector<double>>..  undefined"
   "vectorGuard<Node *>.. undefined"

2. decided to move utils.cpp into header